### PR TITLE
Add Convolutional1dComponent in nnet3 and its test code

### PR DESCRIPTION
**Pull request by freewym**
It passed the test code.

### CODE REVIEW

File: src/nnet3/nnet-test-utils.cc [hunk 1]
```diff
@@ -804,6 +804,18 @@ static void GenerateRandomComponentConfig(std::string *component_type,
       os << "input-dim=" << input_dim << " output-dim=" << output_dim;
       break;
     }
+    case 17: {
+      *component_type = "Convolutional1dComponent";
+      int32 patch_stride = 10, patch_step = 1, patch_dim = 4;
```
Comment (danpovey) [inline src/nnet3/nnet-test-utils.cc]: I think it would be better to test a variety of values here- this may be masking bugs.

File: src/nnet3/nnet-simple-component.h [hunk 2]
```diff
@@ -891,6 +891,131 @@ class NaturalGradientPerElementScaleComponent: public PerElementScaleComponent {
...
+
+  virtual int32 InputDim() const;
+  virtual int32 OutputDim() const;
+
+  virtual std::string Info() const;
+  virtual void InitFromConfig(ConfigLine *cfl);
+  virtual std::string Type() const { return "Convolutional1dComponent"; }
+  virtual int32 Properties() const {
+    return kSimpleComponent|kUpdatableComponent|kBackpropNeedsInput;
+	   // |kBackpropAdds;
```
Comment (danpovey) [inline src/nnet3/nnet-simple-component.h]: Remove this comment.

File: src/nnet3/nnet-simple-component.h [hunk 3]
```diff
@@ -891,6 +891,131 @@ class NaturalGradientPerElementScaleComponent: public PerElementScaleComponent {
...
+  }
+
+  virtual void Propagate(const ComponentPrecomputedIndexes *indexes,
+                 const CuMatrixBase<BaseFloat> &in,
```
Comment (danpovey) [inline src/nnet3/nnet-simple-component.h]: fix indentation

File: src/nnet3/nnet-simple-component.cc [hunk 4]
```diff
@@ -1930,6 +1930,594 @@ void NaturalGradientPerElementScaleComponent::Update(
...
+  }
+
+  if (in_deriv) {
+    std::vector<std::vector<int32> > reversed_column_map;
+    ReverseIndexes(column_map, InputDim(), &reversed_column_map);
+    std::vector<std::vector<int32> > rearranged_column_map;
+    RearrangeIndexes(reversed_column_map, &rearranged_column_map);
+    for (int32 p = 0; p < rearranged_column_map.size(); p++) {
+      CuArray<int32> cu_cols(rearranged_column_map[p]);
+      in_deriv->AddCols(patches_deriv, cu_cols);
```
Comment (danpovey) [inline src/nnet3/nnet-simple-component.cc]: I think maybe this should be CopyCols.  Actually I'm very surprised that the tests passed because this component seems to add to its input and never zero it, yet the properties says that it just sets its input. TestSimpleComponentPropagateProperties() should have failed.  [see output_deriv2].

Comment (danpovey) [inline src/nnet3/nnet-simple-component.cc]: OK, I see now that AddCols is appropriate, because you're calling it multiple times.  But I don't see that you ever zero in_deriv.  Ooh.. I see now.

Comment (freewym) [inline src/nnet3/nnet-simple-component.cc]: It's at the first line of Backprop when doing resizing

File: src/nnet3/nnet-simple-component.cc [hunk 5]
```diff
@@ -1930,6 +1930,594 @@ void NaturalGradientPerElementScaleComponent::Update(
...
+// back propagation function
+void Convolutional1dComponent::Backprop(const std::string &debug_info,
+                                        const ComponentPrecomputedIndexes *indexes,
+                                        const CuMatrixBase<BaseFloat> &in_value,
+                                        const CuMatrixBase<BaseFloat> &out_value,
+                                        const CuMatrixBase<BaseFloat> &out_deriv,
+                                        Component *to_update_in,
+                                        CuMatrixBase<BaseFloat> *in_deriv) const {
+  if (in_deriv)
+    static_cast<CuMatrix<BaseFloat>* >(in_deriv)->Resize(out_deriv.NumRows(), InputDim());
```
Comment (danpovey) [inline src/nnet3/nnet-simple-component.cc]: Ouch!!  You should never static_cast CuMatrixBase to CuMatrix like this!!!  How could this code possibly work if it really pointed to a sub-matrix?  And it will, if it's ever used in a real neural net.    I hope you didn't do this in any previous code.
The code requires that in_deriv already have the correct size.  If you want to zero it you can call SetZero().
But actually, don't call SetZero().  The framework is designed so that if it's more natural to add to the derivative, you just set the flag kBackpropAdds, and it's then the responsibility of the calling code to zero the derivative (if needed).  That way, if the derivative needs to be added to something else, you may be able to avoid redundant copies of matrices.  So add the flag kBackpropAdds, and remove this statement.

Comment (freewym) [inline src/nnet3/nnet-simple-component.cc]: Sorry about that. I did the static_cast because  Backprop() of CNN in nnet2 calls Resize() at the beginning. I will fix it.

**Comment by freewym** [comment]
I fixed the those bugs. Also I removed the call of Resize() at the beginnig of Convolutional1dComponent::Backprop() in nnet2 and changed a line in its nnet-component-test.cc accordingly. Please let me know if there are any problems.
.
Yiming

File: src/nnet3/nnet-test-utils.cc [hunk 6]
```diff
@@ -804,6 +804,24 @@ static void GenerateRandomComponentConfig(std::string *component_type,
       os << "input-dim=" << input_dim << " output-dim=" << output_dim;
       break;
     }
+    case 17: {
+      *component_type = "Convolutional1dComponent";
+      int32 patch_stride = 10 + Rand() % 50, patch_step = 1 + Rand() % 4, 
+	    patch_dim = 4 + Rand () % 5;
+
+      // decrease patch_stride until (patch_stride - patch_dim) % patch_step == 0
+      while ((patch_stride - patch_dim) % patch_step != 0)
```
Comment (freewym) [inline src/nnet3/nnet-test-utils.cc]: Sorry. no while needed. will replace it with a single expression.

Comment (danpovey) [inline src/nnet3/nnet-test-utils.cc]: In nnet2, in_deriv is supposed to be resized by the Backprop routine, that
it why it is typed CuMatrix not CuMatrixBase.  In nnet2, it's different.

Comment (freewym) [inline src/nnet3/nnet-test-utils.cc]: Restored.

Comment (danpovey) [inline src/nnet3/nnet-test-utils.cc]: OK.  Please rebase your changes so that only the initial commit remains.
Dan

Comment (freewym) [inline src/nnet3/nnet-test-utils.cc]: Done. One issue: I left the previous 2 commits (by david and vijay) in the interactive editor when using rebase, so currently there are 3 commits in this pull request. But actually I didn't modify anything in their commits. I am not sure if it is fine for you to merge only mine. If not, I can create a new pull request. Sorry about it.
Yiming

Comment (danpovey) [inline src/nnet3/nnet-test-utils.cc]: You must have done something wrong-- those shouldn't appear there.  See if
you can find someone to help you fix it.  To-morrow is fine.
Meanwhile, you could also try to add a test example in nnet-test-utils.cc
that tests the convolutional component.
Dan

Comment (freewym) [inline src/nnet3/nnet-test-utils.cc]: I fixed that. Now there is only one commit of mine. 
By a test example, did you mean a NN with Convolutional1dComponent, rather than only testing a component? Could you please refer me to a similar example in nnet-test-utils.cc or somewhere else?
Thanks,
Yiming

Comment (danpovey) [inline src/nnet3/nnet-test-utils.cc]: See GenerateConfigSequence and the functions it calls.

Comment (danpovey) [inline src/nnet3/nnet-test-utils.cc]: You can add that to this pull request.

**Comment by danpovey** [comment]
Thanks!  Merging.
I believe in typical convolutional setups there is some kind pooling component (e.g. max-pooling) after each convolutional component.  Perhaps you could look into this, and port any max-pooling code to nnet3 (if necessary), and modify the test config accordingly?
